src/vmm/src/device_manager/mmio.rs

```diff
@@ -110,17 +105,22 @@ impl MMIODeviceManager {
     }
 
     /// Allocates resources for a new device to be added.
-    fn allocate_new_slot(&mut self) -> Result<MMIODeviceInfo> {
-        if self.irq > self.last_irq {
+    fn allocate_new_slot(&mut self, irq_count: u32) -> Result<MMIODeviceInfo> {
```

Comment:
This function does not keep the same abstraction as before this PR.

Previously, it was hiding the implementation details of a slot. A slot was defined as both a irq number & an MMIO base. Not sure this is the best approach for fixing the problem at hand (i.e. having a device that doesn't need interrupts).

To me it looks like a re-design is in order for the interface to still be clean.

Reply:
Possibly beyond my contributor pay grade to comment, as I was working under guidance from @acatangiu here.  

Just to check though, are you saying that you'd preferred `allocate_new_slot` to remain parameterless, and for there to be some other mechanism for determining the number of IRQ's to allocate?

Reply:
Something like that, yap.

It looks a bit like allocate_new_slot is taking care of two things:
- allocate an MMIO slot which is needed for all devices
- allocate an IRQ number

Looking at the code it is a bit hard to fix. The problem seems to be that "slot" is used to define both the irq and the mmio slot. Typically you only call a "slot" the mmio slot.
The expectation would be that the device manager separates the allocation of a slot and of a irq number. Then each device requests irqs & slots as they need them. When a device requests an irq, than irq also needs to be registered against KVM.

You can ignore my comment and I can later add a technical dept issue so we can fix this.

On the other hand, it doesn't look like we ever register more than one irq for every other device. Should we use a bool instead?

Reply:
I see the `mmio space` (composed of `base addr` + `length`) and `used irqs` as resources managed by the `MMIODeviceManager`. For our existing devices these are _hardcoded_ as:
 - `base addr` of current device starts where the previous device `mmio space` ends
 - `mmio space len` is `MMIO_LEN`
 - `used irq(s)` hardcoded to a single IRQ number, next available in the range.

`allocate_new_slot` is the interface through which these resources are reserved for configuration and use by the device being currently set up by the caller.

The way I see it is that this interface can be extended to provide control to the caller on _how many_ of these resources are required for the new device - in this case, the boot timer device doesn't need any IRQs.

Providing more control on what types and how many resources the device needs doesn't leak implementation details for the MMIODeviceManager IMHO. From an interface standpoint, this would ideally allow control over all types and values of resources (base, len, irq count), but I fear that brings a lot of complexity without a practical usecase.

@andreeaflorescu what do you think? Does this make sense?

Reply:
Yap, let's handle this in a separate PR.